src/v/redpanda/application.h

```diff
@@ -259,9 +259,8 @@ class application {
      */
     template<typename Service, typename... Args>
     ss::future<> construct_service(ss::sharded<Service>& s, Args&&... args) {
-        auto f = s.start(std::forward<Args>(args)...);
         _deferred.emplace_back([&s] { s.stop().get(); });
-        return f;
+        return s.start(std::forward<Args>(args)...);
```

Comment:
> If there is an error in the service constructor, the service stop() is
never emplaced in the deferred action thus triggering an assert on
cleanup (_instances.empty() check failed)..

>Swap the order since stop() is idempotent and is a noop if the service
pointer is null, so its safe.

I'm a little confused about this. The second paragraph seems to say its fine to call stop() on a service if it wasn't started. The first seems to say the opposite.

Btw, I wouldn't be surprised if we have lots of services whose stop() method assumes that start() completed.

Reply:
> The first seems to say the opposite.

you mean this sentence "thus triggering an assert on cleanup (_instances.empty() check failed).." says the opposite? If so I meant there that if the sharded object destructor is called without calling stop() (which is possible without this commit), that can trigger an assert.


> Btw, I wouldn't be surprised if we have lots of services whose stop() method assumes that start() completed.

Isn't that the case without this patch too? start(Args..) here only calls the constructor of the object on each shard but the actual start() invocation is explicit in the application code IIUC. That aside I think we should encourage the service implementations to assume that stop() can potentially be called without actually start()-ing, wdyt.

Reply:
>can potentially be called without actually start()-ing, wdyt.

seems like good design to me yeh